File: bindings/python/src/_pyghex/structured/regular/halo_generator.cpp
/*
 * ghex-org
 *
 * Copyright (c) 2014-2023, ETH Zurich
 * All rights reserved.
 *
 * Please, refer to the LICENSE file in the root directory.
 * SPDX-License-Identifier: BSD-3-Clause
 */
#include <array>
#include <tuple>

#include <register_class.hpp>
#include <structured/regular/halo_generator.hpp>

#include <nanobind/nanobind.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/array.h>
#include <nanobind/stl/vector.h>

namespace pyghex
{
namespace structured
{
namespace regular
{

void
register_halo_generator(nanobind::module_& m)
{
    gridtools::for_each<
        gridtools::meta::transform<gridtools::meta::list, halo_generator_specializations>>(
        [&m](auto l)
        {
            using namespace std::string_literals;
            using namespace nanobind::literals;

            using type = gridtools::meta::first<decltype(l)>;
            using dimension = typename type::dimension;
            using array = std::array<int, dimension::value>;
            using halo_array = std::array<int, 2 * dimension::value>;
            using periodic_array = std::array<bool, dimension::value>;
            using box = typename type::box;
            using box2 = typename type::box2;

            auto _halo_generator = register_class<type>(m);
            auto _box = register_class<box>(m);
            auto _box2 = register_class<box2>(m);

            _halo_generator
                .def(nanobind::init<array, array, halo_array, periodic_array>(), "first"_a,
                    "last"_a, "halos"_a, "periodic"_a, "Create a halo generator")
                .def("__call__",
                    [](const type& halo_gen, const typename type::domain_type& domain)
                    {
                        nanobind::list result;
                        for (const auto& halo : halo_gen(domain))
                            result.append(nanobind::cast(halo));
                        return result;
                    });

            _box2.def_prop_ro("local", [](const box2& b) { return b.local(); })
                .def_prop_ro("global_", [](const box2& b) { return b.global(); });

            _box.def_prop_ro("first",
                    [](const box& b)
                    {
                        auto first = b.first();
                        return static_cast<typename decltype(first)::array_type>(first);
                    })
                .def_prop_ro("last",
                    [](const box& b)
                    {
                        auto last = b.last();
                        return static_cast<typename decltype(last)::array_type>(last);
                    });
        });
}

} // namespace regular
} // namespace structured
} // namespace pyghex

File: bindings/python/src/ghex/structured/regular/halo_generator.py
#
# ghex-org
#
# Copyright (c) 2014-2023, ETH Zurich
# All rights reserved.
#
# Please, refer to the LICENSE file in the root directory.
# SPDX-License-Identifier: BSD-3-Clause
#
from __future__ import annotations
from dataclasses import dataclass
from typing import TYPE_CHECKING

from ghex.structured.cartesian_sets import CartesianSet, ProductSet, union
from ghex.util.cpp_wrapper import CppWrapper

if TYPE_CHECKING:
    from typing import Union

    from ghex.structured.regular.domain_descriptor import DomainDescriptor


@dataclass
class HaloContainer:
    local: CartesianSet
    global_: CartesianSet


class HaloGenerator(CppWrapper):
    def __init__(
        self,
        glob_domain_indices: ProductSet,
        halos: tuple[Union[int, tuple[int, int]], ...],
        periodicity: tuple[bool, ...],
    ) -> None:
        assert glob_domain_indices.dim == len(halos)
        assert glob_domain_indices.dim == len(periodicity)

        # canonicalize integer halos, e.g. turn (h0, (h1, h2), h3) into ((h0, h0), (h1, h2), ...)
        halos2 = ((halo, halo) if isinstance(halo, int) else halo for halo in halos)
        flattened_halos = tuple(h for halo in halos2 for h in halo)

        super(HaloGenerator, self).__init__(
            (
                "structured__regular__halo_generator",
                "int",
                glob_domain_indices.dim,
            ),
            glob_domain_indices[tuple(0 for _ in range(glob_domain_indices.dim))],
            glob_domain_indices[tuple(-1 for _ in range(glob_domain_indices.dim))],
            flattened_halos,
            periodicity,
        )

    def __call__(self, domain: DomainDescriptor) -> HaloContainer:
        result = self.__wrapped_call__("__call__", domain)

        local = union(
            *(
                ProductSet.from_coords(tuple(box2.local.first), tuple(box2.local.last))
                for box2 in result
            )
        )
        global_ = union(
            *(
                ProductSet.from_coords(tuple(box2.global_.first), tuple(box2.global_.last))
                for box2 in result
            )
        )
        return HaloContainer(local, global_)
